File: 082_remove_duplicates_from_sorted_list_ii/Solution2/Solution.go
package main

import "fmt"

type ListNode struct {
    Val  int
    Next *ListNode
}

func deleteDuplicates(head *ListNode) *ListNode {
    if head == nil || head.Next == nil { return head }

    dummyHead := &ListNode{-1, nil}
    tail := dummyHead
    for l, r := head, head; l != nil; l = r {
        for r != nil && l.Val == r.Val { r = r.Next }
        if l.Next == r {
            tail.Next = l
            tail = l
            tail.Next = nil
        }
    }

    return dummyHead.Next
}

func main() {
    head := &ListNode{1, nil}
    head.Next = &ListNode{1, nil}
    head.Next.Next = &ListNode{2, nil}
    head.Next.Next.Next = &ListNode{3, nil}
    head.Next.Next.Next.Next = &ListNode{3, nil}

    result := deleteDuplicates(head)
    for result != nil {
        fmt.Println(result.Val)
        result = result.Next
    }
}

File: 082_remove_duplicates_from_sorted_list_ii/Solution2/Solution.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int val) {
        this.val = val;
    }
}

class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null || head.next == null) return head;

        ListNode dummyHead = new ListNode(0);
        ListNode tail = dummyHead;
        for (ListNode l = head, r = head; l != null; l = r) {
            while (r != null && l.val == r.val) r = r.next;
            if (l.next == r) {
                tail.next = l;
                tail = l;
                tail.next = null;
            }
        }

        return dummyHead.next;
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        head.next = new ListNode(1);
        head.next.next = new ListNode(2);
        head.next.next.next = new ListNode(3);
        head.next.next.next.next = new ListNode(3);

        ListNode result = new Solution().deleteDuplicates(head);

        while (result != null) {
            System.out.println(result.val);
            result = result.next;
        }
    }
}
